### PR TITLE
Lock datadir

### LINKED ISSUE #2278: Ensure only one cnd uses a given data dir
## Problem

It is not obvious to the user that if they are running several cnds locally for development purposes, they need to add a `data.dir` parameter to their configuration to ensure the various cnds do not use the same seed and database.

See https://github.com/comit-network/comit-rs/issues/2214#issuecomment-600347964 for example.

## Goal

Making it easy for interested developer to build on top of COMIT.

If a cnd is already started on the given datadir, starting a new cnd should not be possible. This second cnd should die and tell the user that the datadir is already in use.

The user should not have to worry about this lock mechanism. Even if a cnd dies unexpectedly/harshly, the user should not have to do any cleanup.
Be sure to (manually) test with `kill -9`.

## Recommendation

Look if there are any crate that can help with handling of lock files.

### PR DESCRIPTION
**Pull request by rishflab**
Place a lock on the datadir specified in config. cnd aborts with
an error message if the specified datadir is already being used
by another cnd process.

Closes #2278

### CODE REVIEW

**Comment by rishflab** [comment]
I wrapped the logic into a function to improve the error message

```

fn lock_datadir(path: &PathBuf) -> anyhow::Result<()> {
    let file = File::open(path)?;
    match file.try_lock_exclusive() {
        Err(e) => {
            tracing::error!("Could not acquire lock on datadir: {:?}", path);
            Err(anyhow::Error::from(e))
        }
        Ok(_) => Ok(file.lock_exclusive()?),
    }
}
```

When i use the ? operator on this function in main 
```
lock_datadir(&path)?;
```
it doesn't abort the program

am i missing something really obvious?

File: cnd/src/main.rs [hunk 1]
```diff
@@ -52,6 +53,12 @@ fn main() -> anyhow::Result<()> {
 
     crate::trace::init_tracing(settings.logging.level)?;
 
+    let path = &settings.data.dir.to_path_buf();
+
+    let file = std::fs::File::open(&path)?;
+    file.try_lock_exclusive()?;
+    file.lock_exclusive()?;
```
Comment (tcharding) [inline cnd/src/main.rs]: Unless I'm missing something, this is not correct.  You should not be locking the file twice.  Just using `try_lock_exclusive()?`.  Also, can we have this all in a function please, so in main it just reads something like `lock_data_dir(&settings.data.dir)?`.  Thanks

Comment (rishflab) [reply cnd/src/main.rs]: managed to move into a lock_datadir function. It was wasn't working earlier because you need to keep the file object in scope to maintain the lock

Comment (rishflab) [reply cnd/src/main.rs]: also was able to remove the redundant lock statement succesfully

File: config1.toml [hunk 2]
```diff
@@ -0,0 +1,5 @@
+[data]
+dir = "/Users/rishflab/datadir"
```
Comment (thomaseizinger) [inline config1.toml]: I guess these files should not be part of the PR? :D

**Comment by thomaseizinger** [comment]
That `Settings` idea might not be so gooed because the settings and the file lock are actually separate concerns 🤔

File: cnd/src/main.rs [hunk 3]
```diff
@@ -200,3 +203,14 @@ fn dump_config(settings: Settings) -> anyhow::Result<()> {
     println!("{}", serialized);
     Ok(())
 }
+
+fn lock_datadir(path: &PathBuf) -> anyhow::Result<File> {
+    let file = File::open(path)?;
+    match file.try_lock_exclusive() {
+        Err(e) => {
+            tracing::error!("Could not acquire lock on datadir: {:?}", path);
+            Err(anyhow::Error::from(e))
```
Comment (thomaseizinger) [inline cnd/src/main.rs]: Logging and passing on the error is a bit odd, you should be able to use `.context` here to add this information :)

https://docs.rs/anyhow/1.0.27/anyhow/trait.Context.html

Comment (thomaseizinger) [reply cnd/src/main.rs]: So basically doing this:

```rust
File::open(path)?.try_lock_exclusivly().with_context(|| format!("Could not acquire lock on data dir {}", path.display()))?
```

File: cnd/src/main.rs [hunk 4]
```diff
@@ -52,6 +53,8 @@ fn main() -> anyhow::Result<()> {
 
     crate::trace::init_tracing(settings.logging.level)?;
 
+    let _file = lock_datadir(&settings.data.dir)?;
```
[the remark this answers is not in the document]
Comment (rishflab) [reply cnd/src/main.rs]: > That `Settings` idea might not be so gooed because the settings and the file lock are actually separate concerns 🤔

Yeh wasn't completely sure about this. I think we should leave it main for now until there is an obvious and fitting place where to keep it.

File: cnd/src/main.rs (hunk 3 above)
Comment (thomaseizinger) [reply cnd/src/main.rs]: Could also do an extension trait for `PathBuf` to put this code somewhere else and not having it inside `main.rs` :)

File: cnd/src/main.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (rishflab) [reply cnd/src/main.rs]: > If you would change some of the later function calls to use this file, returning it would not be useless :)
> 
> Otherwise, name it something like `_file_lock_to_ensure_we_have_exclusive_access_to_the"_config_until_this_is_dropped`.

The file lock ensures exclusive access to the datadir containing the seed.pem and the sqlite db file not the config.

File: cnd/src/main.rs (hunk 3 above)
Comment (rishflab) [reply cnd/src/main.rs]: > So basically doing this:
> 
> ```rust
> File::open(path)?.try_lock_exclusivly().with_context(|| format!("Could not acquire lock on data dir {}", path.display()))?
> ```

ok ill check his out, seems much cleaner

File: cnd/src/main.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (D4nte) [reply cnd/src/main.rs]: Well, the config also needs to be exclusive otherwise you will have conflict on the ports cnd binds to. But that will be solved as cnd will just crash because it cannot bind to said ports.

Comment (D4nte) [inline cnd/src/main.rs]: Does it print an error on stderr on failure?

Comment (rishflab) [reply cnd/src/main.rs]: No idea tbh. @thomaseizinger  do you know where this error will be printed?

Comment (rishflab) [reply cnd/src/main.rs]: Should I put an exclusive lock on the config file as well or not? You said it will crash anyway due to a port binding failure?

Comment (D4nte) [reply cnd/src/main.rs]: Lock on data dir is enough.
Just try the behaviour locally ;)
